tests/cluster/sql/test_deliverysql.py

```diff
@@ -213,13 +226,88 @@ def test16_upsert_to_table_dummy(self):
         )
 
-        # Validate temp stagning table have been dropped
-        sql_argument = f"""
-        (SELECT * FROM INFORMATION_SCHEMA.TABLES
-        WHERE TABLE_NAME LIKE '{self.table_upsert_name}%') target
-        """
-        table_exists = self.sql_server.load_sql(sql_argument)
-        self.assertEqual(table_exists.count(), 1)
+    def test18_upsert_to_table_with_join_cols_filter(self):
+
+        upsertTableSchema = StructType(
+            [
+                StructField("testid", IntegerType(), True),
+                StructField("testdata", StringType(), True),
+            ]
+        )
+
+        # Write pre upsert test data
+        df_preTest = DataframeCreator.make_partial(
+            schema=upsertTableSchema,
+            columns=["testid", "testdata"],
+            data=[(1, "testdata1"), (2, "testdata2")],
+        )
+        self.sql_server.write_table_by_name(
+            df_source=df_preTest, table_name=self.table_upsert_name, append=False
+        )
+
+        # Write upsert test data with None keys
+        df_upsertTest = DataframeCreator.make_partial(
+            schema=upsertTableSchema,
+            columns=["testid", "testdata"],
+            data=[(2, "newtestdata2"), (3, "testdata3"), (None, "testdata4")],
+        )
+        self.sql_server.upsert_to_table_by_name(
+            df_source=df_upsertTest,
+            table_name=self.table_upsert_name,
+            join_cols=["testid"],
+            filter_join_cols=True,
+        )
+
+        # Validate correct upserted data
+        df_afterUpsert = self.sql_server.read_table_by_name(self.table_upsert_name)
+        expectedData = [(1, "testdata1"), (2, "newtestdata2"), (3, "testdata3")]
+
+        self.assertDataframeMatches(
+            df=df_afterUpsert,
+            expected_data=expectedData,
+        )
+
+    def test19_upsert_to_table_overwrite_empty_target(self):
+
+        upsertTableSchema = StructType(
+            [
+                StructField("testid", IntegerType(), True),
+                StructField("testdata", StringType(), True),
+            ]
+        )
+
+        # Write empty pre upsert test data
+        df_preTest = DataframeCreator.make_partial(
+            schema=upsertTableSchema,
+            columns=["testid", "testdata"],
+            data=[],
+        )
+        self.sql_server.write_table_by_name(
+            df_source=df_preTest, table_name=self.table_upsert_name, append=False
+        )
+
+        # Write upsert test data
+        df_upsertTest = DataframeCreator.make_partial(
+            schema=upsertTableSchema,
+            columns=["testid", "testdata"],
+            data=[(2, "newtestdata2"), (3, "testdata3")],
+        )
+        self.sql_server.upsert_to_table_by_name(
+            df_source=df_upsertTest,
+            table_name=self.table_upsert_name,
+            join_cols=["testid"],
+            filter_join_cols=False,
+            overwrite_if_target_is_empty=False,
```

Comment:
The name is "test19_upsert_to_table_overwrite_empty_target" but the overwrite_if_target_is_empty=false. It this intended?

Reply:
Fixed. But it is actually very hard to fix the overwrite scenario. The end state is the same as a merge, and the temp table is cleaned.

Reply:
I agree.